src/operations/client_bulk_write/results_merger.ts

```diff
@@ -0,0 +1,95 @@
+import { type Document } from '../../bson';
+import { type ClientBulkWriteCursorResponse } from '../../cmap/wire_protocol/responses';
+import {
+  type ClientBulkWriteOptions,
+  type ClientBulkWriteResult,
+  type ClientDeleteResult,
+  type ClientInsertOneResult,
+  type ClientUpdateResult
+} from './common';
+
+/**
+ * Merges client bulk write cursor responses together into a single result.
+ * @internal
+ */
+export class ClientBulkWriteResultsMerger {
+  result: ClientBulkWriteResult;
+  options: ClientBulkWriteOptions;
+
+  /**
+   * Instantiate the merger.
+   * @param options - The options.
+   */
+  constructor(options: ClientBulkWriteOptions) {
+    this.options = options;
+    const baseResult = {
+      insertedCount: 0,
+      upsertedCount: 0,
+      matchedCount: 0,
+      modifiedCount: 0,
+      deletedCount: 0
+    };
+
+    if (options.verboseResults) {
+      this.result = {
+        ...baseResult,
+        insertResults: new Map<number, ClientInsertOneResult>(),
+        updateResults: new Map<number, ClientUpdateResult>(),
+        deleteResults: new Map<number, ClientDeleteResult>()
+      };
+    } else {
+      this.result = baseResult;
+    }
```

Comment:
I think it's out of scope for this ticket because we have a separate TS support ticket in the epic already but I imagine the API actually will look something like:

```typescript
// mongo client
bulkWrite(models, options: ClientBulkWriteOptions & { verboseResults: true }): VerboseResult;
bulkWrite(models, options: ClientBulkWriteOptions & { verboseResults: false }): TerseResult;
bulkWrite(models, options: ClientBulkWriteOptions): TerseResult;
```

In which case, I'd expect not to have the verbose result's keys on the terse result.

Reply:
We can discuss the exact way those types will be defined in that ticket but I think the TS DX of returning two types where they don't share keys is unpleasant because you're forced to narrow vs checking if something is nullish or using `?.` before accessing it. 

The keys could still be present on the `TerseResult` but set to undefined and the TS wouldn't be inaccurate. I feel this is similar to the `in` checks I asked us to change to null checks previously. 

I think it is simpler to write code that does not care about key presence since dot access on a missing key and a key that exists but is equal to undefined end up working the same in practice. My suggestion removes an `else` and the need to spread, so it's "less" work. 

But I don't feel too strongly here, feel free to leave it as is.

Reply:
Just to clarify what I am getting at [checkout the playground here](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAShDOBXANsKBeKBvKBDAXFAHaIC2ARhAE4A0U5A-IfMFQJZEDmUAvgNwAoUJCgA1auQD28CHCSoMsCAEdEbKhAAmAHjkpgAPkHDoAFWoy9CzAHlSbYLoT66AcnKujAgQHofUADlJNABxDVw0ABEADTpTAAtoRCI2SSIoSQAzKGAAd0kc+I5OeCg2UrTkEELoeHjcDU0oAGsIEHg6AEkoUkQWYgaqSVyoXId4vHSAAw4pqABjRPnmqD8170zk+eBU9M1JBOKAfQA3AEYAChPCVkQIAEpCcSopS2dUQU2ibd2ofcOuKdLtcoJlcMgZI8oOYqG95MBPlsdmk-gcioDzldCFJJMgILgiFCYXD9FAAD5iCTSWTvBECL4-FH-dGcIFY+iSXH4wmEKxoLACKBQDTARBUdJYHh4Up8wQ8bwnBpQAAeimZx0xAFkIvEAHRUAn7UgXe5QAxQAAMuoArPdBMrdYxdfBkGx5hALmc7d4-FAAILzd1gYC4ch4qAxOjzAlQABWfTQfWgJGQyAWSxWayz2ZzubzPg2SN+6oxACZ2Ti8QSoXzsILhRBReLsFLcDLaXKFUrqpgS6yTuXtcA9QaiEaTWbLTbvSBHQxna73Z7vYqqFAe6iAf3y2CIQ9BLOnS63R6vXxVv5h+UFpJNNAAKKie8wYgQE7UYXJeC67y+fwAIXvABlUwI2iC980gqDoJgvNC2+ZE9jRDUAGZ2VuB4nipEkPnpIsmWQjE0JBXdIUIYkaXhREEOLQj+2I7FOSrHklCovCaIIrdTgYjkuWrXlaTrIURTFCVW3bNj5QEVcoAALzVOjuIuIcR0NSRjVNc0rVtQRZLnBcTwuUs7QvYhglBSQqHdUpgAKEwBioIYRlyRJ0gcBYY0oKAIHGagfEsrs1zQXtFJONCMO9YBHQMpdjPPX1o3SCBXXsIgImgecMjXFM00WCBllGVzRmgZoiGGMo0GvMANBkIhgGkpVEAUriwouUj9wERB9OPWKTN9aq3wgOqtFBIZSFGdgdi4G870KCJRjYVNX3fNcqGSIA)

Reply:
If I may continue to elaborate on my view of the keys presence/not presence conversation 😅 

Something to keep in mind about typescript is that it only checks that a type is compatible or assignable not that it is complete. So `in` checks and key presence checks is a weak assertion/type narrowing tool given that a function that accepts `{a: number}` can be given `{a: number; b: string}`. At runtime, `b` is present and there's nothing "wrong" with that from the language's perspective. 

If you depend on checking for a key's presence as a way to determine if a type is one thing or the other you will run into problems when you're passed extra keys (if you use `in` the extra keys can be inherited on the prototype). 

I think from a robustness standpoint your goal should be determining if the input is "compatible" with your use case, sort of like "does it quack?" rather than "is it a duck?".

Reply:
Went with Neal's suggestion here.